test/libsolidity/semanticTests/operators/userDefined/calling_all_operators.sol

```diff
@@ -0,0 +1,110 @@
+type Int is int128;
+using {
+    bitor as |, bitand as &, bitxor as ^, bitnot as ~,
+    add as +, sub as -, unsub as -, mul as *, div as /, mod as %,
+    eq as ==, noteq as !=, lt as <, gt as >, leq as <=, geq as >=,
+    not as !
+} for Int;
+
+function uw(Int x) pure returns (int128) {
+    return Int.unwrap(x);
+}
+function w(int128 x) pure returns (Int) {
+    return Int.wrap(x);
+}
+function bitor(Int, Int) pure returns (Int) {
+    return w(10);
+}
+function bitand(Int, Int) pure returns (Int) {
+    return w(11);
+}
+function bitxor(Int, Int) pure returns (Int) {
+    return w(12);
+}
+function bitnot(Int) pure returns (Int) {
+    return w(13);
+}
+function add(Int x, Int) pure returns (Int) {
+    return w(uw(x) + 10);
+}
+function sub(Int, Int) pure returns (Int) {
+    return w(15);
+}
+function unsub(Int) pure returns (Int) {
+    return w(16);
+}
+function mul(Int, Int) pure returns (Int) {
+    return w(17);
+}
+function div(Int, Int) pure returns (Int) {
+    return w(18);
+}
+function mod(Int, Int) pure returns (Int) {
+    return w(19);
+}
+function eq(Int x, Int) pure returns (bool) {
+    return uw(x) == 1;
+}
+function noteq(Int x, Int) pure returns (bool) {
+    return uw(x) == 2;
+}
+function lt(Int x, Int) pure returns (bool) {
+    return uw(x) < 10;
+}
+function gt(Int x, Int) pure returns (bool) {
+    return uw(x) > 10;
+}
+function leq(Int x, Int) pure returns (bool) {
+    return uw(x) <= 10;
+}
+function geq(Int x, Int) pure returns (bool) {
+    return uw(x) >= 10;
+}
+function not(Int x) pure returns (bool) {
+    return uw(x) == 0;
+}
+
+contract C {
+    function test_bitor() public pure returns (Int) { return w(1) | w(2); }
+    function test_bitand() public pure returns (Int) { return w(1) & w(2); }
+    function test_bitxor() public pure returns (Int) { return w(1) ^ w(2); }
+    function test_bitnot() public pure returns (Int) { return ~w(1); }
+    function test_add(int128 x) public pure returns (Int) { return w(x) + w(2); }
+    function test_sub() public pure returns (Int) { return w(1) - w(2); }
+    function test_unsub() public pure returns (Int) { return -w(1); }
+    function test_mul() public pure returns (Int) { return w(1) * w(2); }
+    function test_div() public pure returns (Int) { return w(1) / w(2); }
+    function test_mod() public pure returns (Int) { return w(1) % w(2); }
+    function test_eq(int128 x) public pure returns (bool) { return w(x) == w(2); }
+    function test_neq(int128 x) public pure returns (bool) { return w(x) != w(2); }
+    function test_lt(int128 x) public pure returns (bool) { return w(x) < w(2); }
+    function test_gt(int128 x) public pure returns (bool) { return w(x) > w(2); }
+    function test_leq(int128 x) public pure returns (bool) { return w(x) <= w(2); }
+    function test_geq(int128 x) public pure returns (bool) { return w(x) >= w(2); }
+    function test_not() public pure returns (bool) { return !w(1); }
```

Comment:
I'm confused about the semantics given that this implemetation returns a boolean yet should return `Int` based on the statement, "! returns the same type it takes rather than bool. Should it be returning bool instead?", [here](https://forum.soliditylang.org/t/feature-preview-user-defined-operators/1435).

Also, I'd expect the "not" operator to only apply to user defined types whose underlying type can also use the operator. In this case, I would expect a similar error to trying `!int128(1)`: `error[4907]: TypeError: Unary operator ! cannot be applied to type int128.`

Reply:
I tried this locally and observed the behavior explained in the feature preview, so I'm not sure why this is passing?
```
Error: Wrong return parameters in operator definition. The function "not" needs to return exactly one value of type Int to be used for the operator !.
  --> test.sol:63:34:
   |
63 | function not(Int x) pure returns (bool) {
   |                                  ^^^^
```
My concern regarding the second point still stands

Reply:
There's some confusion here so let me explain what happened:
- I posted the thread. At that point `!` was still returning the argument type. I wasn't sure if that's a good design so I included a question about it.
- I also posted this question here (https://github.com/ethereum/solidity/pull/13790#discussion_r1072798198). Daniel answered with a good point that it should be returning `bool`.
- Later I applied this change and pushed the change here. Now in the PR you have it returning `bool` but the thread still has the binary from before that change.
- Another change here compared to the binaries in the thread is that `<<`, `>>` and `**` can no longer be redefined (see https://github.com/ethereum/solidity/pull/13790#discussion_r1072807608 for discussion).
- I'm going to update the binary in the thread and say what changed, either today or on Monday. I want to keep it up to date but I also think that posting there after every single change in the PR would be both tedious for me and confusing to people so I'm going to do it only after a few changes accumulate. Hopefully there won't be any more changes at this point anyway. For the purpose of high-level discussion and user feedback I think that should be enough granularity.
- I did not expect someone to actually come to the PR and review it (and thanks for looking at it! - feel free to comment on the code). If you're looking at the PR directly it's best to use the current binary from the CI. The ones on the forum link to what was in CI a few runs ago. You can get to the current runs via the "Details" links on CI checks below. The jobs that give you static binaries are : `b_ubu_static`, `b_win`, `b_osx` and `b_ems`.

Reply:
> Also, I'd expect the "not" operator to only apply to user defined types whose underlying type can also use the operator. In this case, I would expect a similar error to trying `!int128(1)`: `error[4907]: TypeError: Unary operator ! cannot be applied to type int128.`

That's actually what happens, You can only use these operators on user-defined types. This test may be confusing because for every operator definition it also has a `test_xyz()` helper function that calls it and these functions take raw integers and wrap them in the custom `Int` type just before using the operator. The operations inside them are actually performed on the custom type, not on `int128`. This might not be apparent because the test hides the wrapping in a another helper function: `w()`.

We use this style with test helpers because in the test expectations we can only call external functions. Operators are internal and cannot be used in expectations directly.

I think I can actually rewrite this test to be more straightforward. For exampie, we could have the helpers use `Int` rather than `int128`. We could also group the operations in one helper. I'll push a change with that in a moment.

Reply:
Please check out the updated test. I hope it's clearer now.

Reply:
Thanks for the clarification and pointers. That's my for bad using an outdated binary and reviewing recent changes. I was able to download an up-to-date binary.

My point was related to code like the following:
```
type Int is int128;
using {
    not as !
} for Int;

function not(Int x) pure returns (bool) {
    return Int.unwrap(x) == 0;
}

function a() public pure returns (bool) {
      require(!Int.wrap(1));
      if (!Int.wrap(1)) {
          return false;
      }
      return !Int.wrap(1);
  }
```
If I were to try this with the underlying type, `int128`, I would receive a type error indicating that the unary not operator is not available for this type. I don't see why wrapping it in a type alias should change that behavior. What the function `not` is doing is arbitrary and isn't related to logical negation as that operation cannot be performed on it's underlying type. It would make since to instead define a helper function instead:
```
function is_zero(Int x) pure returns (bool) {
     return Int.unwrap(x) == 0;
}
function b() public pure returns (bool) {
      require(is_zero(Int.wrap(1)));
      if (is_zero(Int.wrap(1))) {
          return false;
      }
      return is_zero(Int.wrap(1));
}
```
My suggestion would be to only allow the not operator for user defined types whose underlying type is `bool`. Otherwise, the not operator will just be used to check an arbitrary condition on a value type and no longer mean logical negation. I think enforcing the latter option would be best for clarity and be in line with the existing semantics of the not operator.

Reply:
Ah, now I see what you meant. You make a good point. I'm still on the fence with that though so let's wait for some input from @ekpyron or @chriseth. I think you should also raise that point in the thread so other users can chime in. That's exactly the kind of a design issue I was hoping for people to point out and state their opinion on before we release it.

If I had to argue for keeping it as is, I'd say that the underlying type of the UDVT should be treated as an implementation detail. You might want to implement a type with some boolean-like behavior but backed by something other than `bool`. So in that sense allowing `!` is ok because there are UDVTs for which it semantically makes sense. And we can't really prevent people from doing all unreasonable things with UDVTs in all cases. You could define operators with completely crazy semantics and we can't stop you :) This is a trade-off and the question is if restricting it does more good than harm.

On the other hand, one thing that reasonates with me in your argument is that I'm pretty sure people will use this feature to implement truthy/falsy expressions, like `!x` instead of `x == 0`. I think this style makes conditions less readable and is also error prone (`0` is false but sometimes gets forgotten when your check is meant to catch `null`). Still, that's just my personal opinion and there's no `null` in Solidity so there's much less danger in doing that.

Another good argument for disallowing it would be that we did just that with other operators where the design has some uncertainties, like `**`, `++` or `&&`. This way we can still enable them later in a non-breaking release when we have a better idea how it will mesh with other features we have in the pipeline now (especially generics). Allowing them now and disallowing later would be breaking.

Reply:
In the end we decided to disallow `!` in the initial release.